### PR TITLE
Support GEOSEARCH and GEOSEARCHSTORE commands

### PR DESCRIPTION
**Pull request by AvitalFineRedis**
Support GEOSEARCH and GEOSEARCHSTORE commands.
Also created GeoSearchParam class to support all the params for both commands.

### CODE REVIEW

**Comment by sazzad16** [comment]
@yangbodong22011 @dengliming @gkorland @mina-asham 

Should we name `geosearch` or `geoSearch`?

**Comment by yangbodong22011** [comment]
> @yangbodong22011 @dengliming @gkorland @mina-asham
> 
> Should we name `geosearch` or `geoSearch`?

If you follow the old style (georadius, georadiusStore), it should be:
`geosearch`  
`geosearchStore`

**Comment by sazzad16** [comment]
Also, what about the name of method for `geosearchstore` with `storedist`?

- `geosearchStoreStoreDist` - according to Redis doc (and existing pattern)
- `geosearchStoreDist` - alters existing pattern
- `geosearchStoreWithDist` - meaningful but doesn't match Redis doc

**Comment by yangbodong22011** [comment]
> Also, what about the name of method for `geosearchstore` with `storedist`?
> 
> * `geosearchStoreStoreDist` - according to Redis doc (and existing pattern)
> * `geosearchStoreDist` - alters existing pattern
> * `geosearchStoreWithDist` - meaningful but doesn't match Redis doc

🤣 i vote for `geosearchStoreStoreDist `

**Review by sazzad16 — DISMISSED** [review]

**Review by sazzad16 — CHANGES_REQUESTED** [review]
Please use

- `geosearchStore`
- `geosearchStoreStoreDist`

in proper places.

Also, separate `GeoRadiusParam` and `GeoSearchParam`. There would be some codes twice but, IMO, it's a better trade off compared to introducing new patterns.

**Review by yangbodong22011 — COMMENTED** [review]
It seems that there is a lack of adding corresponding commands in `MultiKeyPipelineBase.java` and `PipelineBase.java`

File: src/main/java/redis/clients/jedis/MultiKeyPipelineBase.java [hunk 1]
```diff
@@ -952,6 +952,42 @@
     return getResponse(BuilderFactory.LONG);
   }
 
+  @Override
+  public Response<Long> geosearchStore(String dest, String src, String member, double radius, GeoUnit unit) {
```
Comment (yangbodong22011) [inline src/main/java/redis/clients/jedis/MultiKeyPipelineBase.java:956]: need add
```
public void geosearchStore(byte[] dest, byte[] src, byte[] member, double radius, GeoUnit unit)
```

in pipeline, binary command and string command are in same file.

File: src/main/java/redis/clients/jedis/BinaryJedisCluster.java [hunk 2]
```diff
@@ -2259,6 +2259,112 @@ public Double execute(Jedis connection) {
...
+  }
+
+  @Override
+  public long geosearchStore(byte[] dest, byte[] src, byte[] member, double width, double height, GeoUnit unit) {
+    return new JedisClusterCommand<Long>(connectionHandler, maxAttempts, maxTotalRetriesDuration) {
+      @Override
+      public Long execute(Jedis connection) {
+        return connection.geosearchStore(dest, src, member, width, height, unit);
+      }
+    }.runBinary(src);}
```
Comment (yangbodong22011) [inline src/main/java/redis/clients/jedis/BinaryJedisCluster.java:2339]: `} `Need to wrap...The following three methods are also.

Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BinaryJedisCluster.java:2339]:
```suggestion
    }.runBinary(src);
  }
```

File: src/main/java/redis/clients/jedis/BinaryJedisCluster.java [hunk 3]
```diff
@@ -2259,6 +2259,112 @@ public Double execute(Jedis connection) {
...
+
+  @Override
+  public long geosearchStore(byte[] dest, byte[] src, double longitude, double latitude, double width, double height, GeoUnit unit) {
+    return new JedisClusterCommand<Long>(connectionHandler, maxAttempts, maxTotalRetriesDuration) {
+      @Override
+      public Long execute(Jedis connection) {
+        return connection.geosearchStore(dest, src, longitude, latitude, width, height, unit);
+      }
+    }.runBinary(src);}
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BinaryJedisCluster.java:2348]:
```suggestion
    }.runBinary(src);
  }
```

File: src/main/java/redis/clients/jedis/BinaryJedisCluster.java [hunk 4]
```diff
@@ -2259,6 +2259,112 @@ public Double execute(Jedis connection) {
...
+
+  @Override
+  public long geosearchStore(byte[] dest, byte[] src, GeoSearchParam params) {
+    return new JedisClusterCommand<Long>(connectionHandler, maxAttempts, maxTotalRetriesDuration) {
+      @Override
+      public Long execute(Jedis connection) {
+        return connection.geosearchStore(dest, src, params);
+      }
+    }.runBinary(src);}
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BinaryJedisCluster.java:2357]:
```suggestion
    }.runBinary(src);
  }
```

File: src/main/java/redis/clients/jedis/BinaryJedisCluster.java [hunk 5]
```diff
@@ -2259,6 +2259,112 @@ public Double execute(Jedis connection) {
...
+
+  @Override
+  public long geosearchStoreStoreDist(byte[] dest, byte[] src, GeoSearchParam params) {
+    return new JedisClusterCommand<Long>(connectionHandler, maxAttempts, maxTotalRetriesDuration) {
+      @Override
+      public Long execute(Jedis connection) {
+        return connection.geosearchStoreStoreDist(dest, src, params);
+      }
+    }.runBinary(src);}
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BinaryJedisCluster.java:2366]:
```suggestion
    }.runBinary(src);
  }
```

File: src/main/java/redis/clients/jedis/JedisCluster.java [hunk 6]
```diff
@@ -2429,6 +2429,119 @@ public Double execute(Jedis connection) {
...
+  }
+
+  @Override
+  public long geosearchStore(String dest, String src, double longitude, double latitude, double width, double height, GeoUnit unit) {
+    return new JedisClusterCommand<Long>(connectionHandler, maxAttempts, maxTotalRetriesDuration) {
+      @Override
+      public Long execute(Jedis connection) {
+        return connection.geosearchStore(dest, src, longitude, latitude, width, height, unit);
+      }
+    }.run(src);  }
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisCluster.java:2523]:
```suggestion
    }.run(src);
  }
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]
For MultiKey commands (geosearchStore...), `.run(src);` -> `.run(2, dest, src);`.
Similar for binary commands.

File: src/main/java/redis/clients/jedis/JedisCluster.java [hunk 7]
```diff
@@ -2429,6 +2429,119 @@ public Double execute(Jedis connection) {
...
+  }
+
+  @Override
+  public long geosearchStore(String dest, String src, double longitude, double latitude, double width, double height, GeoUnit unit) {
+    return new JedisClusterCommand<Long>(connectionHandler, maxAttempts, maxTotalRetriesDuration) {
+      @Override
+      public Long execute(Jedis connection) {
+        return connection.geosearchStore(dest, src, longitude, latitude, width, height, unit);
+      }
+    }.run(2, dest, src);  }
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisCluster.java:2523]:
```suggestion
    }.run(2, dest, src);
  }
```

**Review by sazzad16 — APPROVED** [review]

**Review by yangbodong22011 — APPROVED** [review]

**Review by dengliming — APPROVED** [review]

**Pull request closed without merging**